### DISCUSSION TITLE
Merging two layouts

### DISCUSSION THREAD

**Discussion by senekor** [zellij-org/zellij#3164] [Ideas]
I have an idea for a feature, but I'm not 100% sure if this is already possible somehow or if it's even the right approach to my problem / use case.

What I would like to do is merge user-layouts and project-layouts. For example, I have my own user-layout in `~/.config/zellij/layouts/ide.kdl` where the main thing right now is a tab that opens helix automatically. Other people may like to use vim instead, or maybe I want to have a second tab with btop open in the future... you get the point, these are personal preferences.

Now the other thing I love to use zellij for is project-layouts. For example, I may have a layout in a git repository under `dev/zellij.kdl` with tabs / panes to run frontend & backend, render documentation and so on. This is very project specific. (Then I usually have a convenient `just` recipe to start zellij with that layout.)

Now, I would like to combine these two in some way. High-level: "Please start zellij, with this base layout and load these additional tabs".

The base layout should be the user layout and be as powerful as normal zellij layouts. (ui settings, keybinds etc. supported). The project specific additional tabs don't necessarily need to be that powerful, it can just be loading a number of additional tabs with panes and commands in them to run. I'm saying this because I think it could be difficult to define how ui settings and keybinds from two layouts are correctly merged.

I'm wondering what other people think about this? Is it already possible and I don't know how? Is this a valid use case for zellij to try to support? Is there a better approach to achieve my goal?

**Comment by imsnif** [comment]
Hey, cool use case! I definitely want Zellij to make it as easy as possible.

Right now, you can *kind of* do this by opening the project layouts as new tabs with the `new-tab` action (eg. `zellij action new-tab --layout project-layout.kdl`), but this only works if these layouts have just one tab.

I want to make it possible to open a new layout from within Zellij regardless of how many tabs it has. Essentially for `zellij --layout file.kdl` to detect that you're inside a Zellij session, and "do the right thing" (essentially, open the new layout as new tab(s)). I'm afraid I can't provide an ETA for this though...

Another option is to use the `session-manager` and view these new layouts as sessions rather than additional tabs, then you can switch between them without exiting Zellij. Starting next version, the session-manager will be able to start new sessions with specific layouts, also allowing you to choose the cwd for each such layout.

Hope this helps!

**Reply by senekor** [reply]
Thank you for the swift response!

> I'm afraid I can't provide an ETA for this though...

If we find something to implement that's sufficiently well specified to support such a use case, I might have a shot at implementing it myself.

---

> Right now, you can kind of do this by opening the project layouts as new tabs with the `new-tab` action (eg. `zellij action new-tab --layout project-layout.kdl`), but this only works if these layouts have just one tab.

I think this is almost powerful enough to support my use case. I don't think it's a problem that only one tab can be loaded. I can just have a folder `dev/zellij-tabs/*.kdl` in my projects and load all of them one-by-one.

The manual way to do that would be:
- Open zellij normally, with your user-layout
- Run project script to load additional tabs (e.g. `just load-zellij-tabs` or whatever)

This is pretty good already, but I'd love to reduce that to one step. I think that would require the ability to control a session from outside of it. Pseudo-script:

```sh
session_name="$(zellij create-session)" # <- I don't think this command exists yet
for tab in $(ls dev/zellij-tabs/)
do
    # I don't think the "session" flag exists yet
    zellij action --session "$session_name" new-tab --layout "$tab"
done
zellij attach "$session_name"
```

As far as I can tell there are two new things that need to be implemented for this:
- create new session without attaching to it (e.g. `zellij create-session`)
- send actions to sessions without being attached to them (e.g. `zellij action --session <SESSION_NAME>`)

What do you think about this? It seems to me this approch would potentially unlock many more use cases we're not even thinking about here. It would enable almost arbitrarily controlling zellij sessions from scripts.

**Reply by senekor** [reply]
I got it working! The missing piece was `close_on_exit`. With that one can run arbitrary scripts within a zellij layout and not mess up the layout itself by leaving behind a useless pane. So my current setup is:

- The user-layout contains one pane that looks like this:
  ```
  pane {
    command "just"
    args "_load-zellij-tabs"
    close_on_exit true
  }
  ```
- Project specific tabs can be loaded with an arbitrary script executed by `just _load-zellij-tabs`. This can be a single tab, or several hardcoded ones or even a dynamic number of them with a loop over directory contents etc.

Limitations to this approach:
1. In order to use this, users must have the special tab-loading pane in their regular user-layout. I think that's fine. If you put it in your dotfiles, it's a one-and-done thing. Users can make separate user-layouts for themselves where some of them do laod project-tabs and some don't.
2. The convention of `just _load-zellij-tabs` is arbitrary. Maybe some other convention would be better.
3. All the project tabs are opened to the right of the user tabs. I would prefer them to be on the left. But it looks like [tab reordering is about to land](https://github.com/zellij-org/zellij/pull/3047). Lucky me!

I think I will be happy with this approach for myself. But it's not really an intuitive or obvious workflow. The features it uses don't advertise themselves as fulfilling this purpose. It's basically abusing a `close_on_exit`-pane to dynamically generate a layout from a script. Maybe this is a candidate for cookbook-style documentation? I would be happy to contribute it, if you like.
